### PR TITLE
update reader in dygraph

### PR DESCRIPTION
**Pull request by yghstill**
- YOLOv3:
  - Dygraph new Dataloader:
single p40: avg 0.51s pre batch
8 card p40: avg 0.91s pre batch

  - Dygraph old Reader：
single p40: avg 0.57s pre batch
8 card p40: avg 0.96s pre batch

  - Static graph Reader:
single p40: avg 0.47s pre batch
8 card p40: avg 0.98s pre batch

- Mask-RCNN-FPN：
  - Dygraph new Dataloader:
single p40: avg 0.32s pre batch
8 card p40: avg 0.58s pre batch
  - Dygraph old Reader：
single p40: avg 0.72s pre batch
8 card p40: avg 1.39s pre batch

### CODE REVIEW

File: ppdet/data/sampler.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+from paddle import fluid
+from paddle.io import BatchSampler
+from paddle.fluid.layers import collective
+from paddle.fluid.dygraph.parallel import ParallelEnv, ParallelStrategy
+
+_parallel_context_initialized = False
+
+
+class DistributedBatchSampler(BatchSampler):
```
Comment (heavengate) [inline ppdet/data/sampler.py:18]: 这个应该是有实现的`paddle.io.DistributedBatchSampler`

File: ppdet/data/source/coco.py [hunk 2]
```diff
@@ -80,7 +50,7 @@ def load_roidb_and_cname2cid(self, with_background=True):
         # when with_background = True, mapping category to classid, like:
         #   background:0, first_class:1, second_class:2, ...
         catid2clsid = dict({
-            catid: i + int(with_background)
+            catid: i + int(self.with_background)
```
Comment (heavengate) [inline ppdet/data/source/coco.py:53]: 这里`self.`的已经没有了

File: tools/eval.py [hunk 3]
```diff
@@ -19,7 +19,7 @@
 from ppdet.utils.check import check_gpu, check_version, check_config
 from ppdet.utils.cli import ArgsParser
 from ppdet.utils.eval_utils import get_infer_results, eval_results
-from ppdet.data.reader import create_reader
+#from ppdet.data.reader import create_reader
```
Comment (heavengate) [inline tools/eval.py:22]: delete

File: configs/_base_/datasets/coco.yml [hunk 4]
```diff
@@ -2,17 +2,17 @@ metric: COCO
 num_classes: 80
 
 TrainDataset:
-  !COCODataSet
+    name: COCODataset
     image_dir: train2017
     anno_path: annotations/instances_train2017.json
-    dataset_dir: dataset/coco
+    dataset_dir: /paddle/data/coco #dataset/coco
```
Comment (heavengate) [inline configs/_base_/datasets/coco.yml:8]: 这里改过来吧

File: configs/_base_/datasets/coco.yml [hunk 5]
```diff
@@ -2,17 +2,17 @@ metric: COCO
 num_classes: 80
 
 TrainDataset:
-  !COCODataSet
+    name: COCODataset
```
Comment (heavengate) [inline configs/_base_/datasets/coco.yml:5]: name方式制定能create出来么？这个后面可能需要跟models配置统一一下看看使用name还是用！

File: ppdet/data/sampler.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+import time
+import math
+import socket
+import contextlib
+import numpy as np
+
+from paddle import fluid
+from paddle.io import BatchSampler
+from paddle.fluid.layers import collective
+from paddle.fluid.dygraph.parallel import ParallelEnv, ParallelStrategy
```
Comment (jerrywgz) [inline ppdet/data/sampler.py:13]: use paddle.distributed

File: ppdet/data/sampler.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    def __len__(self):
+        num_samples = self.num_samples
+        num_samples += int(not self.drop_last) * (self.batch_size - 1)
+        return num_samples // self.batch_size
+
+    def set_epoch(self, epoch):
+        self.epoch = epoch
+
+
+def _all_gather(x, nranks, ring_id=0, use_calc_stream=True):
```
Comment (jerrywgz) [inline ppdet/data/sampler.py:89]: is _all_gather used?

File: ppdet/data/source/__init__.py [hunk 8]
```diff
@@ -13,9 +13,9 @@
 # limitations under the License.
 
 from . import coco
-from . import voc
-from . import widerface
+#from . import voc
```
Comment (jerrywgz) [inline ppdet/data/source/__init__.py:16]: add TODO

File: configs/_base_/datasets/coco.yml (hunk 5 above)
Comment (yghstill) [reply configs/_base_/datasets/coco.yml:5]: 继续使用！，thx。

File: configs/_base_/datasets/coco.yml (hunk 4 above)
Comment (yghstill) [reply configs/_base_/datasets/coco.yml:8]: done,thx.

File: ppdet/data/sampler.py (hunk 6 above)
Comment (yghstill) [reply ppdet/data/sampler.py:13]: done, `ParallelStrategy`待优化。

File: ppdet/data/sampler.py (hunk 1 above)
Comment (yghstill) [reply ppdet/data/sampler.py:18]: 这个后续会补充增强。

File: ppdet/data/sampler.py (hunk 7 above)
Comment (yghstill) [reply ppdet/data/sampler.py:89]: deleted，thx

File: ppdet/data/source/__init__.py (hunk 8 above)
Comment (yghstill) [reply ppdet/data/source/__init__.py:16]: done

File: ppdet/data/source/coco.py (hunk 2 above)
Comment (yghstill) [reply ppdet/data/source/coco.py:53]: done，thx

File: tools/eval.py (hunk 3 above)
Comment (yghstill) [reply tools/eval.py:22]: done，thx

File: ppdet/data/transform/batch_operators.py [hunk 9]
```diff
@@ -55,7 +56,7 @@ def __init__(self, pad_to_stride=0, use_padded_im_info=True, pad_gt=False):
         self.use_padded_im_info = use_padded_im_info
         self.pad_gt = pad_gt
 
-    def __call__(self, samples, context=None):
+    def __call__(self, samples):
```
Comment (qingqing01) [inline ppdet/data/transform/batch_operators.py:59]: context其实在一些场景下还可能用到，虽然暂且没用到，先去除也行吧

Comment (yghstill) [reply ppdet/data/transform/batch_operators.py:59]: ok

**Comment by qingqing01** [comment]
LGTM. @heavengate @jerrywgz Please see again.

**Review by heavengate — APPROVED** [review]

**Review by jerrywgz — APPROVED** [review]
LGTM, use_prefetch and with_background in Reader could be removed in next PR.

**Pull request merged**